apps/homescreen/js/pages.js

```diff
@@ -0,0 +1,174 @@
+/* global PagesStore, IconsHelper */
+'use strict';
+
+/**
+ * The size in pixels of the icons in the pinned pages.
+ */
+const PAGES_ICON_SIZE = 30;
+
+(function(exports) {
+
+  function Pages() {
+    // Element references
+    this.panels = document.getElementById('panels');
+    this.pages = document.getElementById('pages');
+    this.shadow = document.querySelector('#pages-panel > .shadow');
+    this.scrollable = document.querySelector('#pages-panel > .scrollable');
+
+    // Scroll behaviour
+    this.scrolled = false;
+
+    // Track when the first pinned page has been added
+    this.firstPinnedPage = true;
+
+    // Signal handlers
+    this.pages.addEventListener('click', this);
+    this.pages.addEventListener('contextmenu', this);
+    this.scrollable.addEventListener('scroll', this);
+    window.addEventListener('hashchange', this);
+
+    // Initialise and populate pinned pages
+    this.pagesStore = new PagesStore('places');
+    this.pagesStore.init().then(() => {
+      // Triggered when pages are added and updated.
+      document.addEventListener('places-set', (e) => {
+        var id = e.detail.id;
+        this.pagesStore.get(id).then((page) => {
+          if (!page.data.pinned) {
+            return;
+          }
+
+          for (var child of this.pages.children) {
+            if (child.dataset.id === id) {
+              this.updatePinnedPage(child, page.data);
+              return;
+            }
+          }
+
+          // A new page was pinned.
+          this.addPinnedPage(page.data);
+        });
+      });
+
+      document.addEventListener('places-removed', (e) => {
+        var id = e.detail.id;
+        for (var child of this.pages.children) {
+          if (child.dataset.id === id) {
+            this.pages.removeChild(child);
+            return;
+          }
+        }
+      });
+
+      document.addEventListener('places-cleared', () => {
+        for (var child of this.pages.children) {
+          this.pages.removeChild(child);
+        }
+      });
+    },
+    (e) => {
+      console.error('Error initialising pinned pages', e);
+    }).then(() => {
+      return this.pagesStore.getAll().then((pages) => {
+        for (var page of pages) {
+          this.addPinnedPage(page.data);
+        }
+      }, (e) => {
+        console.error('Error getting pinned pages', e);
+      });
+    });
+  }
+
+  Pages.prototype = {
+    updatePinnedPage: function(card, page) {
+      card.title = page.title;
+      card.dataset.id = page.url;
+      card.style.order = -Math.round(page.pinTime / 1000);
+
+      setTimeout(function(card, page) {
+        var background = {};
+        if (page.screenshot) {
+          background.src = URL.createObjectURL(page.screenshot);
+        }
+        if (page.themeColor) {
+          background.themeColor = page.themeColor;
+        }
+        card.background = background;
+
+        IconsHelper.getIconBlob(page.url, PAGES_ICON_SIZE, page)
```

Comment:
Shouldn't we use the shared/homescreen/icon here, as we do in app_chrome? That will use the GaiaIconRenderer for applying the same effects in all places.